### DISCUSSION TITLE
Creation of user_ids

### DISCUSSION THREAD

**Discussion by ACPK** [plausible/analytics#568] [Self-Hosted Support]
The site makes use of the preview salt in order to generate user_ids.

For data scientists that are trying to understand the raw data, would it be possible to get a quick description on how user_ids are generated?

````
previous_user_id = salts[:previous] && generate_user_id(conn, params, salts[:previous])
event = struct(Plausible.ClickhouseEvent, event_attrs)
session_id = Plausible.Session.Store.on_event(event, previous_user_id)
````

**Comment by ukutaht** [comment]
Sure, you can find an explanation here: https://plausible.io/data-policy#how-we-count-unique-users-without-cookies

**Reply by ACPK** [reply]
@ukutaht - Thank you for the quick reply and hope your year is off to a great start! I had read the post but am still confused.

The post mentions _"Old salts are deleted to avoid the possibility of linking visitor information from one day to the next."_ 

Is the previous salt not an old salt (i.e., ````salts[:previous])````) and being used to link visitor information from one day to the next?

**Reply by ukutaht** [reply]
We keep the latest 2 salts around so that session data doesn't get broken at midnight abritrarily. When salts are rotated, the `salts[:previous]` part ensures that active sessions are carried over and not split in half. Old salts are deleted once they have served their purpose meaning within 48 hours of their creation.
